docs/proposals/synchronous-garbage-collection.md

```diff
@@ -0,0 +1,201 @@
+
+## OwnerReference
+
+```go
+OwnerReference {
+     ...
+     // If true, AND if the owner has the "CollectingGarbage" finalizer, then the owner cannot be deleted from the key-value store until this reference is removed.
+     // Defaults to false.
+     // To set this field, a user needs "update" and "delete" permission of the owner, otherwise 422 (Unprocessable Entity) will be returned.
+     BlockOwnerDeletion *bool
+}
+```
+
+The initial draft of the proposal did not include this field and it had a security loophole: a user who is only authorized to update one resource can set ownerReference to block the synchronous GC of other resources. Requiring users to explicitly set `BlockOwnerDeletion` allows the master to properly authorize the request.
+
+## DeleteOptions
+
+```go
+DeleteOptions {
+  …
+  // Whether and how garbage collection will be performed.
+  // Defaults to GarbageCollectionDefault
+  GarbageCollectionPolicy GarbageCollectionPolicy
+}
+
+type GarabgeCollectionPolicy string
+
+const (
+    // Respects the existing garbage collection related finalizers on the object and the default garbage collection policy of the resource.
+    GarbageCollectionDefault GarbageCollectionPolicy = "Default"
+    // Orphans the dependents
+    GarbageCollectionOrphan GarbageCollectionPolicy = "Orphan"
+    // Deletes the object from the key-value store, the garbage collector will delete the dependents in the background.
+    GarbageCollectionAsynchronous GarbageCollectionPolicy = "AsynchronousGarbageCollection"
+    // The object exists in the key-value store until the garbage collector deletes all the dependents whose ownerReference.blockOwnerDeletion=true from the key-value store.
+    // API sever will put the "CollectingGarbage" finalizer on the object, and sets its deletionTimestamp.
+    // This policy is cascading, i.e., the dependents will be deleted with GarbageCollectionSynchronous.
+    GarbageCollectionSynchronous GarbageCollectionPolicy = "SynchronousGarbageCollection"
+)
+```
+
+`DeleteOptions.OrphanDependents *bool` will be removed. We decided not to add a `DeleteOptions.SynchronousGC *bool`, because together with `DeleteOptions.OrphanDependents` it results in 9 possible combinations and is thus confusing. This is a breaking change, so the new DeleteOptions will live in `api.k8s.io` group with version `v1`, as proposed in [#33900](https://github.com/kubernetes/kubernetes/pull/33900).
+
+The conversion rules are described in the following table:
+
+| 1.5                           | pre 1.4/1.4              |
+|-------------------------------|--------------------------|
+| GarbageCollectionDefault      | OrphanDependents==nil    |
+| GarbageCollectionOrphan       | *OrphanDependents==true  |
+| GarbageCollectionAsynchronous | *OrphanDependents==false |
+| GarbageCollectionSynchronous  | N/A                      |
```

Comment:
Do you prefer calling the two GC modes as Asynchronous/Synchronous or Background/Foreground?